onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -261,6 +300,32 @@ Status UnpackTensorWithExternalData(const ONNX_NAMESPACE::TensorProto& tensor,
                                           reinterpret_cast<unsigned char*>(p_data));
 }
 
+#define DEFINE_INT4_UNPACK_TENSOR_WITH_EXT_DATA_IMPL(INT4_TYPE)                                                      \
+  template <>                                                                                                        \
+  Status UnpackTensorWithExternalData<INT4_TYPE>(const ONNX_NAMESPACE::TensorProto& tensor,                          \
+                                                 const ORTCHAR_T* tensor_proto_dir, size_t expected_num_elements,    \
+                                                 /*out*/ INT4_TYPE* p_data) {                                        \
+    static_assert(std::is_trivially_copyable<INT4_TYPE>::value, "T must be trivially copyable");                     \
+                                                                                                                     \
+    ORT_RETURN_IF(nullptr == p_data, "nullptr == p_data");                                                           \
+    std::vector<uint8_t> unpacked_tensor;                                                                            \
+    ORT_RETURN_IF_ERROR(ReadExternalDataForTensor(tensor, tensor_proto_dir, unpacked_tensor));                       \
+                                                                                                                     \
+    size_t num_packed_pairs = INT4_TYPE::CalcNumInt4Pairs(expected_num_elements);                                    \
+    ORT_RETURN_IF_NOT(num_packed_pairs == unpacked_tensor.size(), "Unexpected number of packed int4 pairs");         \
+                                                                                                                     \
+    gsl::span<const INT4_TYPE> src_span = gsl::make_span(reinterpret_cast<const INT4_TYPE*>(unpacked_tensor.data()), \
+                                                         num_packed_pairs);                                          \
+    gsl::span<INT4_TYPE> dst_span = gsl::make_span(p_data, expected_num_elements);                                   \
```

Comment:
`gsl::make_span(p_data, expected_num_elements)`

should the span length be num_packed_pairs?

is there much benefit to using spans here if they're just provided to memcpy?